pkgs/development/tools/build-managers/bazel/common.nix

```diff
@@ -0,0 +1,563 @@
+{ version
+, sha256
+, src_deps
+, javac11
+, stdenv, callPackage, lib, fetchurl, fetchFromGitHub
+, runCommand, runCommandCC, makeWrapper, recurseIntoAttrs
+# this package (through the fixpoint glass)
+, bazel
+, lr, xe, zip, unzip, bash, writeCBin, coreutils
+, which, gawk, gnused, gnutar, gnugrep, gzip, findutils
+# updater
+, python27, python3, writeScript
+# Apple dependencies
+, cctools, libcxx, CoreFoundation, CoreServices, Foundation
+# Allow to independently override the jdks used to build and run respectively
+, buildJdk, runJdk
+, buildJdkName
+, runtimeShell
+# Downstream packages for tests
+, bazel-watcher
+# Always assume all markers valid (this is needed because we remove markers; they are non-deterministic).
+# Also, don't clean up environment variables (so that NIX_ environment variables are passed to compilers).
+, enableNixHacks ? false
+, gcc-unwrapped
+, autoPatchelfHook
+, file
+, substituteAll
+, writeTextFile
+}:
+
+let
+  inherit version src_deps;
+
+  src = fetchurl {
+    url = "https://github.com/bazelbuild/bazel/releases/download/${version}/bazel-${version}-dist.zip";
+    inherit sha256;
+  };
+
+  # Update with `eval $(nix-build -A bazel.updater)`,
+  # then add new dependencies from the dict in ./src-deps.json as required.
+  srcDeps = lib.attrsets.attrValues srcDepsSet;
+  srcDepsSet =
+    let
+      srcs = (builtins.fromJSON (builtins.readFile src_deps));
+      toFetchurl = d: lib.attrsets.nameValuePair d.name (fetchurl {
+        urls = d.urls;
+        sha256 = d.sha256;
+        });
+        in builtins.listToAttrs (map toFetchurl [
+      srcs.desugar_jdk_libs
+      srcs.io_bazel_skydoc
+      srcs.bazel_skylib
+      srcs.io_bazel_rules_sass
+      srcs.platforms
+      (if stdenv.hostPlatform.isDarwin
+       then srcs."java_tools_javac11_darwin-${javac11}.zip"
+       else srcs."java_tools_javac11_linux-${javac11}.zip")
+      srcs."coverage_output_generator-v2.0.zip"
```

Comment:
How stable are these names across versions? This likely has to be change from one version to another. It might be best to have a helper function around this and pass them in?

It would probably be fine to do that in the `default.nix`. If things are largely the same we can just use a function to generate this based on the sources file of each bazel version.

Reply:
I'm not quite sure to be honest, I can look over the commit history/bazel 0.26

Reply:
It seems like javac11 changes, but it's definitely not the only thing that's changing. Maybe this should be generated from the sources file, but I'm not quite sure how to implement that yet.

Reply:
What changes? If you give me an example I can try to guide you.

Reply:
https://github.com/NixOS/nixpkgs/commit/ba327a54b67732a715cc4237f5de20411b635319

https://github.com/NixOS/nixpkgs/commit/950da6263f7f861246ac284e45f51ff0073e9d8d

https://github.com/NixOS/nixpkgs/commit/67904dccbb0dfc01c9f8d242db6302cbdfcdb195

https://github.com/NixOS/nixpkgs/commit/7e5b2e74b7ee31e26fb3ced45ef74758d682505a

Looks mostly like android tools, javac, and the bazel toolchain

Reply:
Ok, so my approach would be to add an argument to the common function that takes a function that can be used to filter the dependencies.

e.g. add something like this to the `common.nix`:
```
let
  mkSrcs = srcSpec: srcsFn:
    let
       toFetchurl = d: lib.attrsets.nameValuePair d.name (fetchurl {
           urls = d.urls;
           sha256 = d.sha256;
       });
      srcs = (builtins.fromJSON (builtins.readFile srcSpec));
      selectedSrcs = depsFn srcs;
    in
      in builtins.listToAttrs (map (dep: toFetchurl) selectedSrcs)):
in
…
srcDepsSet = mkSrcs srcDepsJson srcDepsSelector;
…
```

and in the `default.nix`:
```
in {
	bazel_23 = common {
	  srcDepsJson = ./src-deps.json;
	  srcDepsSelector = srcs:
	  [
	    (if stdenv.hostPlatform.isDarwin then
	      srcs."java_tools_javac11_darwin-v4.0.zip"
	    else
	      srcs.""java_tools_javac11_linux-v4.0.zip")

	    srcs."coverage_output_generator-v1.0.zip"
	    srcs."android_tools_pkg-0.8.tar.gz"
	    srcs."0.27.1.tar.gz"
	    srcs.rules_pkg
	    srcs.rules_cc
	    srcs.rules_java
	  ]
	};
}

```

Reply:
Ah, this is really clean! Heading to bed now, but will implement this tomorrow